### PR TITLE
Run nond tests in parallel batches

### PR DESCRIPTION
**Pull request by marusak**
[fedora-31/firefox](https://logs.cockpit-project.org/logs/pull-14422-20200729-061421-ab92649c-fedora-31-firefox/log)  # TESTS PASSED [2760s on 2-ci-srv-06, 137 serial tests: 0: 2466s, 1: 2354s]
[debian-stable](https://logs.cockpit-project.org/logs/pull-14422-20200729-061421-ab92649c-debian-stable/log) # TESTS PASSED [1700s on 3-ci-srv-02, 137 serial tests: 0: 1334s, 1: 1665s]
[fedora-31](https://logs.cockpit-project.org/logs/pull-14422-20200729-061421-ab92649c-fedora-31/log) # 1 TESTS FAILED [2764s on 4-ci-srv-06, 137 serial tests: 0: 2477s, 1: 1975s]
[fedora-32](https://logs.cockpit-project.org/logs/pull-14422-20200729-061421-ab92649c-fedora-32/log) # 1 TESTS FAILED [3589s on 1-ci-srv-03, 137 serial tests: 0: 3583s, 1: 1490s]

### CODE REVIEW

**Comment by marusak** [comment]
Waaay to many tests are failing. Mostly seems like tests that provision VMs, but they are destructive by design.

File: test/common/run-tests [hunk 1]
```diff
@@ -278,27 +301,30 @@ def run(opts, image):
...
 
                     # sometimes our global machine gets messed up; also, tests that time out don't run cleanup handlers
                     # restart it to avoid an unbounded number of test retries and follow-up errors
                     if not opts.machine and (poll_result == 124 or (retry_reason and b"test harness" in retry_reason)):
                         # try hard to keep the test output consistent
-                        testlib.MachineCase.kill_global_machine() # TODO?
-                        testlib.MachineCase.get_global_machine()
+                        testlib.MachineCase.kill_global_machine(test.serial_machine)
+                        # TODO is the ssh adn browser address the same?
+                        testlib.MachineCase.get_global_machine(restrict=not opts.enable_network, id=test.serial_machine)
```
Comment (marusak) [inline test/common/run-tests:314]: @martinpitt Here we can destroy machine and get a new one. But we  generate commands with ssh_address and web_address before that. Will we get the same ports?

**Comment by marusak** [comment]
hmm, `--list` shows only count of nond tests, but does not list their names.
Also wrong rebase when dropping the one problematic commit so fails on cleanup.

**Comment by marusak** [comment]
First results in. Two queues seem to reduce the time nicely, but now the destructive tests take the longest for sure. 2 jobs for 170 destructive tests and 2 jobs for 137 nond tests does not seem completely fair (destructive have a bit more overhead with always preparing and destroying VMs). It feel like 5 jobs split to 3 and 2 could provide interesting results. WDYT @martinpitt ?

I want to also count time that destructive tests take so we can have proper time stats. And will also write now simple test that will test more batches.

Other than that it is ready for the first review.

File: test/common/run-tests (hunk 1 above)
Comment (martinpitt) [reply test/common/run-tests:324]: You are right that this is a potential race condition if you run more than one `run-tests` at the same time on a machine. That never happens in CI though, and hopefully not too much on developer machines either. It's fine to mark this as a FIXME, but we've had this for a long time already.

**Comment by martinpitt** [comment]
@marusak : That's why I didn't add a time for the destructive tests back then. :-) The idea was that the total test time indicates how long they take *if* they take longer than the serial ones. But the serial ones have dominated the test time so far, so clearly the parallel ones had enough resources.

It's great to see that now the parallel ones dominate again! They [roughly take 1½ the time of a serial batch now](https://logs-https-cockpit.apps.ci.centos.org/logs/pull-14422-20200728-192406-91f371db-fedora-32/log) , so indeed it seems that `$TEST_JOBS=5` might be the sweet spot. I deployed that to e2e and CentOS CI machines now. Want to give this a try? If that works out as expected, I'll update cockpituous/playbooks accordingly.

File: test/common/run-tests [hunk 2]
```diff
@@ -285,7 +285,7 @@ def run(opts, image):
                     # restart it to avoid an unbounded number of test retries and follow-up errors
                     if not opts.machine and (poll_result == 124 or (retry_reason and b"test harness" in retry_reason)):
                         # try hard to keep the test output consistent
-                        testlib.MachineCase.kill_global_machine()
+                        testlib.MachineCase.kill_global_machine() # TODO?
```
Comment (martinpitt) [inline test/common/run-tests]: Let's replace this with `# FIXME: May change the ports when a parallel run-tests is running"

File: test/common/run-tests [hunk 3]
```diff
@@ -160,11 +161,11 @@ def run(opts, image):
     test_loader = unittest.TestLoader()
     parallel_tests = []
     serial_tests = []
+    batch_tests = {}
```
Comment (martinpitt) [inline test/common/run-tests:164]: This could do with a comment about how the map looks like.

File: test/common/run-tests [hunk 4]
```diff
@@ -228,35 +229,67 @@ def run(opts, image):
...
+    if serial_tests and opts.list:
+        # Just build one batch for listing
+        batch_tests["list"] = {"working": None, "tests": [], "time": 0}
+        for test in serial_tests:
+            test.serial_machine = "list"
+            batch_tests["list"]["tests"].append(test)
+
     if serial_tests and not opts.list:
         if opts.machine:
+            batch_tests["theM"] = {"working": None, "tests": [], "time": 0}
```
Comment (martinpitt) [inline test/common/run-tests:241]: "theM" :grinning: More seriously, the indices could just be integers? That would also avoid the str(i) thingies

File: test/common/run-tests [hunk 5]
```diff
@@ -228,35 +229,67 @@ def run(opts, image):
...
+                batch_tests[str(i)] = {"working": None, "tests": [], "time": 0}
+                m = testlib.MachineCase.get_global_machine(restrict=not opts.enable_network, id=str(i))
+                ssh_address = "{0}:{1}".format(m.ssh_address,
+                                               m.ssh_port)
+                web_address = "{0}:{1}".format(m.web_address,
+                                               m.web_port)
+
+                batch = serial_tests[batch_size * i : batch_size * i + batch_size]
+                if i == opts.batches - 1: # Last machine needs to resolve the rest
+                    batch = serial_tests[batch_size * i : ]
```
Comment (martinpitt) [inline test/common/run-tests:265]: Let's please do that with an "else:" branch.

File: test/common/run-tests [hunk 6]
```diff
@@ -376,7 +376,7 @@ def main():
     parser.add_argument('--exclude', action="append", default=[], metavar="TestClass.testName",
                         help="Exclude test (exact match only); can be specified multiple times")
     parser.add_argument('-b', '--batches', type=int,
-                        default=os.environ.get("TEST_BATCHES", 1), help="Number of concurrent batches of nondestructive tests")
+                        default=os.environ.get("TEST_BATCHES", 2), help="Number of concurrent batches of nondestructive tests")
```
Comment (martinpitt) [inline test/common/run-tests:379]: How about defaulting to max(TEST_JOBS//2, 1) ? With the current measurements and evolution of serial vs. parallel tests that feels like DTRT.

 I.e. determine TEST_JOBS outside of `add_argument()`, and use it for --jobs and --batches.

File: test/common/run-tests (hunk 4 above)
Comment (marusak) [reply test/common/run-tests:241]: as in "the Machine" :)
But yeah, I guess we can  just use integers, I used `str(i)` as I saw that you used `"1"` in global machines. I'll update it to use strings.

**Comment by marusak** [comment]
[centos-8-stream](https://logs.cockpit-project.org/logs/pull-14422-20200729-073645-075e08f5-centos-8-stream/log) took rather long # 2 TESTS FAILED [5812s on 4-ci-srv-03, 137 serial tests: 0: 5802s, 1: 2086s]
Other than that it seems reasonably nicely balanced. @martinpitt do you think it is something to be worried about?

**Comment by martinpitt** [comment]
@marusak: Worried in general, yes -- there were 7 retries of Machines tests, and some of them took way too long. But not worried in the context of this PR. I'd just like to retry, and make sure that this at least *can* get all the tests to green.

**Comment by marusak** [comment]
Ok, now [centos-8-stream](https://logs.cockpit-project.org/logs/pull-14422-20200729-094154-075e08f5-centos-8-stream/log) worked nicely `# TESTS PASSED [2822s on 2-ci-srv-01, 138 serial tests: 0: 1972s, 1: 1752s]`
Just one debian failure about coredump.

**Pull request merged**